### Imports

In [441]:
from csv         import reader
from math        import sqrt, floor
from collections import Counter
from statistics  import fmean, stdev, correlation

### Data preparation

In [442]:
with open("./data/computer-sales.csv", newline="") as csvfile:
    spamreader = reader(csvfile, delimiter=",", quotechar="|")
    
    # Converting '_csv.reader' to 'list'
    raw_sales_data = list(spamreader)

    print("First 5 rows:")
    for row in raw_sales_data[:5]:
        print(row)


col_len = len(raw_sales_data[0])
row_len = len(raw_sales_data)
print(f"\nRows: {row_len}, Columns: {col_len}")


First 5 rows:
['Sale ID', 'Company', 'Contact', 'Contact Sex', 'Contact Age', 'Contact City', 'Contact State', 'Product Company', 'Product ID', 'Product Type', 'Sale Price', 'Our Cost', 'Shipping Cost', 'Profit', 'Lead Source', 'Sale Month', 'Sale Year']
['1', 'PNT Designs', 'Paul Thomas', 'M', '43', 'Carnegie', 'OH', 'HP', 'M01-F0024', 'Desktop', '479.99', '315.50', '21.10', '143.39', 'Website', 'January', '2018']
['2', 'Ace Manufacturing', 'Margo Simms', 'F', '37', 'Larimer', 'WV', 'HP', 'GT13-0024', 'Desktop', '1249.99', '998.00', '21.10', '230.89', 'Flyer 4', 'January', '2018']
['3', "Sammie's", 'Sam Stine', 'M', '26', 'Pittsburgh', 'PA', 'Dell', 'I3670', 'Desktop', '649.99', '510.25', '21.10', '118.64', 'Website', 'February', '2018']
['4', 'One Moe Time', 'Moe Eggert', 'M', '35', 'St. Clair', 'PA', 'Dell', 'I3593', 'Laptop', '399.99', '310.50', '17.40', '72.09', 'Website', 'March', '2018']

Rows: 40, Columns: 17


In [443]:
# Getting the header
sales_data = {i: [] for i in raw_sales_data[0]}

In [444]:
# Accessing nth key in the dictionary
def get_nth_key(dictionary, n=0):
    if n < 0: # negative indexing
        n += len(dictionary)
    for i, key in enumerate(dictionary):
        if i == n:
            return key
    raise IndexError("Dictionary index out of range")

idx = 1
print(f"Getting index-{idx}: {get_nth_key(sales_data, idx)}")

Getting index-1: Company


In [445]:
# Making sales_data_list to a dictionary
for col in range(col_len):
    for row in range(1, row_len):
        val = raw_sales_data[row][col]
        sales_data[get_nth_key(sales_data, col)].append(val)

In [446]:
# Accessing nth key's value in the dictionary
def get_nth_value(dictionary, idx):
    return dictionary[get_nth_key(dictionary, idx)]

idx = 1
print(get_nth_value(sales_data, idx))

# NOTE: Alternative
# sales_data["Company"]

['PNT Designs', 'Ace Manufacturing', "Sammie's", 'One Moe Time', 'Get Going Gym', 'ANX Trucking', 'Samms Grooming', 'Roberts Produce', 'Klondike Dairy', 'Jones Manufacturing', 'James Cycles', 'Weight Stalkers', 'Case Solutions', "Doug's House", 'Helms Manufacturing', 'Collins Advertising', 'Owens & Sons', 'Samms Grooming', 'Roberts Produce', 'ANX Trucking', 'Case Solutions', "Doug's House", 'PNT Designs', 'Ace Manufacturing', 'Samms Grooming', 'Roberts Produce', 'Klondike Dairy', 'One Moe Time', 'Get Going Gym', 'Jones Manufacturing', 'James Cycles', 'Weight Stalkers', 'Collins Advertising', 'Owens & Sons', 'Samms Grooming', 'ANX Trucking', 'Case Solutions', "Doug's House", 'One Moe Time']


In [447]:
# Converting columns to appropriate datatypes
print("Before:")
## -> int
print(f"Sale ID:", type(sales_data["Sale ID"][0])) 
print(f"Contact Age:", type(sales_data["Contact Age"][0])) 
print(f"Sale Year:", type(sales_data["Sale Year"][0])) 
## -> float
print(f"Sale Price:", type(sales_data["Sale Price"][0])) 
print(f"Our Cost:", type(sales_data["Our Cost"][0]))
print(f"Shipping Cost:", type(sales_data["Shipping Cost"][0])) 
print(f"Profit:", type(sales_data["Profit"][0])) 

dtype_columns = {
    "int": ["Sale ID", "Contact Age", "Sale Year"],
    "float": ["Sale Price", "Our Cost", "Shipping Cost", "Profit"]
}

for dtype in dtype_columns:
    for col in dtype_columns[dtype]:
        if dtype == "int":
            sales_data[col] = [int(i) for i in sales_data[col]]
        elif dtype == "float":
            sales_data[col] = [float(i) for i in sales_data[col]]

print("\nAfter:")
## -> int
print(f"Sale ID:", type(sales_data["Sale ID"][0])) 
print(f"Contact Age:", type(sales_data["Contact Age"][0])) 
print(f"Sale Year:", type(sales_data["Sale Year"][0])) 
## -> float
print(f"Sale Price:", type(sales_data["Sale Price"][0])) 
print(f"Our Cost:", type(sales_data["Our Cost"][0]))
print(f"Shipping Cost:", type(sales_data["Shipping Cost"][0])) 
print(f"Profit:", type(sales_data["Profit"][0]))      

Before:
Sale ID: <class 'str'>
Contact Age: <class 'str'>
Sale Year: <class 'str'>
Sale Price: <class 'str'>
Our Cost: <class 'str'>
Shipping Cost: <class 'str'>
Profit: <class 'str'>

After:
Sale ID: <class 'int'>
Contact Age: <class 'int'>
Sale Year: <class 'int'>
Sale Price: <class 'float'>
Our Cost: <class 'float'>
Shipping Cost: <class 'float'>
Profit: <class 'float'>


### Accessing the data

#### To acess the rows and columns

In [448]:
for k, v in sales_data.items():
    print(f"{k}: {v}")

Sale ID: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39]
Company: ['PNT Designs', 'Ace Manufacturing', "Sammie's", 'One Moe Time', 'Get Going Gym', 'ANX Trucking', 'Samms Grooming', 'Roberts Produce', 'Klondike Dairy', 'Jones Manufacturing', 'James Cycles', 'Weight Stalkers', 'Case Solutions', "Doug's House", 'Helms Manufacturing', 'Collins Advertising', 'Owens & Sons', 'Samms Grooming', 'Roberts Produce', 'ANX Trucking', 'Case Solutions', "Doug's House", 'PNT Designs', 'Ace Manufacturing', 'Samms Grooming', 'Roberts Produce', 'Klondike Dairy', 'One Moe Time', 'Get Going Gym', 'Jones Manufacturing', 'James Cycles', 'Weight Stalkers', 'Collins Advertising', 'Owens & Sons', 'Samms Grooming', 'ANX Trucking', 'Case Solutions', "Doug's House", 'One Moe Time']
Contact: ['Paul Thomas', 'Margo Simms', 'Sam Stine', 'Moe Eggert', 'Jessica Elk', 'Sally Struthers', 'Michelle Samms', 'Mick Roberts', 

#### To access the data row wise

In [449]:
def get_nth_row(dictionary, idx):
    return [v[idx-1] for v in dictionary.values()]

# Getting 1st row or 0th index row
get_nth_row(sales_data, 1) 

[1,
 'PNT Designs',
 'Paul Thomas',
 'M',
 43,
 'Carnegie',
 'OH',
 'HP',
 'M01-F0024',
 'Desktop',
 479.99,
 315.5,
 21.1,
 143.39,
 'Website',
 'January',
 2018]

#### To access the data column wise

In [450]:
sales_data["Sale ID"]

[1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39]

#### Acessing `Profit` column based on ANY column

In [451]:
def get_key_profit_dict(dictionary):
    freq_dict = Counter(dictionary)
    profict_dict =  {i: [] for i in freq_dict.keys()}

    for i in profict_dict:
        for j in range(row_len-1):
            if dictionary[j] == i:
                profict_dict[i].append(sales_data["Profit"][j])

    return profict_dict

sex_profit_list = get_key_profit_dict(sales_data["Contact Sex"])
print(f"Sex Profit List: {sex_profit_list}")

Sex Profit List: {'M': [143.39, 118.64, 72.09, 146.69, 122.34, 143.39, 180.34, 122.34, 118.64, 146.69, 72.09, 143.39, 180.34, 122.34, 143.09, 143.09, 98.09, 143.39, 180.34, 143.39, 180.34, 72.09], 'F': [230.89, 98.09, 230.89, 180.34, 230.89, 72.09, 98.09, 146.69, 122.34, 143.09, 118.64, 180.34, 230.89, 72.09, 98.09, 146.69, 122.34]}


### Analysis

### `What is the mean profit for ANY columns

In [452]:
def get_profit_mean_category_dict(dictionary, ordered=False):
    category_profit_list = get_key_profit_dict(dictionary)
    category_profit_mean =  {i: round(fmean(category_profit_list[i]), 2)  for i in category_profit_list.keys()}
    if ordered:
        return dict(sorted(category_profit_mean.items(), key=lambda i: i[1], reverse=True))
    else:
        return category_profit_mean

In [453]:
target_columns = [#"Company",
                  #"Contact",
                  "Contact Sex",
                  #"Contact Age",
                  #"Contact City",
                  "Contact State",
                  "Product Company",
                  #"Product ID",
                  "Product Type",
                  "Lead Source",
                  "Sale Month",
                  "Sale Year"]

for col in target_columns:
    print(f"{col.upper()}:\n {get_profit_mean_category_dict(sales_data[col], ordered=True)}")

CONTACT SEX:
 {'F': 148.38, 'M': 133.48}
CONTACT STATE:
 {'WV': 165.19, 'OH': 151.17, 'NY': 131.34, 'PA': 130.46}
PRODUCT COMPANY:
 {'ASUS': 167.92, 'HP': 156.37, 'Apple': 146.69, 'Lenovo': 122.34, 'Dell': 89.55}
PRODUCT TYPE:
 {'Desktop': 166.37, 'Tablet': 146.69, 'Laptop': 125.04}
LEAD SOURCE:
 {'Flyer 3': 180.34, 'Flyer 1': 163.51, 'Flyer 2': 158.07, 'Flyer 4': 138.49, 'Website': 131.12, 'Email': 123.2}
SALE MONTH:
 {'November': 156.4, 'January': 154.02, 'April': 153.13, 'February': 149.21, 'August': 148.69, 'May': 144.04, 'December': 137.37, 'July': 134.51, 'March': 93.19}
SALE YEAR:
 {2018: 152.38, 2020: 138.97, 2019: 127.41}


In [454]:
for col in target_columns:
    print(f"\n{col} Data:")
    print(f"\t{Counter(sales_data[col])}")
    print(f"\t{col} Profit List: {get_key_profit_dict(sales_data[col])}")
    print(f"\t{col} Profit Mean: {get_profit_mean_category_dict(sales_data[col], ordered=True)}")


Contact Sex Data:
	Counter({'M': 22, 'F': 17})
	Contact Sex Profit List: {'M': [143.39, 118.64, 72.09, 146.69, 122.34, 143.39, 180.34, 122.34, 118.64, 146.69, 72.09, 143.39, 180.34, 122.34, 143.09, 143.09, 98.09, 143.39, 180.34, 143.39, 180.34, 72.09], 'F': [230.89, 98.09, 230.89, 180.34, 230.89, 72.09, 98.09, 146.69, 122.34, 143.09, 118.64, 180.34, 230.89, 72.09, 98.09, 146.69, 122.34]}
	Contact Sex Profit Mean: {'F': 148.38, 'M': 133.48}

Contact State Data:
	Counter({'PA': 19, 'OH': 11, 'NY': 5, 'WV': 4})
	Contact State Profit List: {'OH': [143.39, 180.34, 146.69, 122.34, 230.89, 146.69, 98.09, 122.34, 143.09, 230.89, 98.09], 'WV': [230.89, 143.39, 143.09, 143.39], 'PA': [118.64, 72.09, 98.09, 230.89, 180.34, 122.34, 118.64, 72.09, 72.09, 143.39, 180.34, 143.09, 98.09, 180.34, 180.34, 72.09, 143.39, 180.34, 72.09], 'NY': [146.69, 122.34, 118.64, 146.69, 122.34]}
	Contact State Profit Mean: {'WV': 165.19, 'OH': 151.17, 'NY': 131.34, 'PA': 130.46}

Product Company Data:
	Counter({'HP

- Assume that `higher profit mean` means `more pofitable category`

- Why we're just assuming it?
    - Because (one of such example):
    
        - The mean profit provides an average figure, but it doesn't reflect the volume of transactions. 
        - A category with a high average profit but very few sales may contribute less overall profit than a category with a slightly lower average profit but many more sales. 
        - For a more comprehensive view, look at both the average profit per sale and the total profit from all sales in that category.

#### Custom Statistics Functions

In [455]:
def custom_mean(data):
    return sum(data) / len(data)

custom_mean([1, 2, 3, 4, 5])

3.0

In [456]:
def custom_stdev(data):
    x_bar = custom_mean(data)
    # To mitigate the error when len(data) = 1
    try:
        return sqrt(sum((x-x_bar)**2 for x in data) / (len(data)-1))
    except:
        return sqrt(sum((x-x_bar)**2 for x in data) / 1)

custom_stdev([1.5, 2.5, 2.5, 2.75, 3.25, 4.75])

1.0810874155219827

In [457]:
def custom_coefficient_variation(data):
    return round(custom_stdev(data) / custom_mean(data), 4)

miles = [3, 4, 4.5, 3.5]
km = [4.828, 6.437, 7.242, 5.632]

cv_miles = custom_coefficient_variation(miles)
print(f"Coefficient Variation of Miles: {cv_miles}")
cv_km = custom_coefficient_variation(km)
print(f"Coefficient Variation of Km: {cv_km}")

if (cv_miles == cv_km):
    print(f"Coefficient Variation (Both): {cv_miles*100}%")

Coefficient Variation of Miles: 0.1721
Coefficient Variation of Km: 0.1721
Coefficient Variation (Both): 17.21%


In [458]:
def custom_covariance(x, y):
    x_bar = custom_mean(x)
    y_bar = custom_mean(y)
    cov = sum((x-x_bar)*(y-y_bar) for x, y in zip(x,y)) / (len(x) - 1)
    return round(cov, 2)

market_cap = [1532, 1488, 1343, 928, 615]
earnings = [58, 35, 75, 41, 17]

cov_MC_E = custom_covariance(market_cap, earnings)
print(f"Covariance: {cov_MC_E}")

if cov_MC_E > 0:
    print("Positive covariance (increases / moving together)")
elif cov_MC_E < 0:
    print("Negative covariance (decreases / moving in opposite direction)")
else: # cov_MC_E = 0 
    print("Independent")

Covariance: 5803.2
Positive covariance (increases / moving together)


In [459]:
def custom_correlation(x, y):
    S_x = custom_stdev(x)
    S_y = custom_stdev(y)
    return round(custom_covariance(x,y) / (S_x*S_y), 4)

r_MC_E = custom_correlation(market_cap, earnings)
print(f"Correlation (Pearson's): {r_MC_E}")

if round(r_MC_E) == 1:
    print("Perfect positive linear relationship")
elif round(r_MC_E) == -1:
    print("Perfect negative linear relationship")
elif round(r_MC_E) == 0:
    print("No linear relationship")

Correlation (Pearson's): 0.6601
Perfect positive linear relationship


In [460]:
def get_profit_stat_category_dict(dictionary, stat, custom=True):
    category_profit_list = get_key_profit_dict(dictionary)
    if stat.lower() == "mean":
        if custom:
            return  {i: round(custom_mean(category_profit_list[i]), 2)  for i in category_profit_list.keys()}
        else:
            return  {i: round(fmean(category_profit_list[i]), 2)  for i in category_profit_list.keys()}
    elif stat.lower() == "stdev":
        if custom:
            return  {i: round(custom_stdev(category_profit_list[i]), 2)  for i in category_profit_list.keys()}
        else:
            return  {i: round(stdev(category_profit_list[i]), 2)  for i in category_profit_list.keys()}
    elif stat.lower() == "cv": # Only custom func is available
        return  {i: round(custom_coefficient_variation(category_profit_list[i]), 2)  for i in category_profit_list.keys()}

print(get_profit_stat_category_dict(sales_data["Contact Sex"], stat="mean"))
print(get_profit_stat_category_dict(sales_data["Contact Sex"], stat="mean", custom=False))
print(get_profit_stat_category_dict(sales_data["Contact Sex"], stat="stdev"))
print(get_profit_stat_category_dict(sales_data["Contact Sex"], stat="stdev", custom=False))
print(get_profit_stat_category_dict(sales_data["Contact Sex"], stat="cv"))

{'M': 133.48, 'F': 148.38}
{'M': 133.48, 'F': 148.38}
{'M': 33.22, 'F': 56.33}
{'M': 33.22, 'F': 56.33}
{'M': 0.25, 'F': 0.38}


In [461]:
for col in target_columns:
    print(f"\n{col} Data:")
    print(f"\t{dict(Counter(sales_data[col]))}")
    print(f"\t{col} Profit List: {get_key_profit_dict(sales_data[col])}")
    print(f"\t{col} Profit Mean: {get_profit_stat_category_dict(sales_data[col], stat="mean")}")
    print(f"\t{col} Standard Deviation: {get_profit_stat_category_dict(sales_data[col], stat="stdev")}")
    print(f"\t{col} Coefficient Variation: {get_profit_stat_category_dict(sales_data[col], stat="cv")}")


Contact Sex Data:
	{'M': 22, 'F': 17}
	Contact Sex Profit List: {'M': [143.39, 118.64, 72.09, 146.69, 122.34, 143.39, 180.34, 122.34, 118.64, 146.69, 72.09, 143.39, 180.34, 122.34, 143.09, 143.09, 98.09, 143.39, 180.34, 143.39, 180.34, 72.09], 'F': [230.89, 98.09, 230.89, 180.34, 230.89, 72.09, 98.09, 146.69, 122.34, 143.09, 118.64, 180.34, 230.89, 72.09, 98.09, 146.69, 122.34]}
	Contact Sex Profit Mean: {'M': 133.48, 'F': 148.38}
	Contact Sex Standard Deviation: {'M': 33.22, 'F': 56.33}
	Contact Sex Coefficient Variation: {'M': 0.25, 'F': 0.38}

Contact State Data:
	{'OH': 11, 'WV': 4, 'PA': 19, 'NY': 5}
	Contact State Profit List: {'OH': [143.39, 180.34, 146.69, 122.34, 230.89, 146.69, 98.09, 122.34, 143.09, 230.89, 98.09], 'WV': [230.89, 143.39, 143.09, 143.39], 'PA': [118.64, 72.09, 98.09, 230.89, 180.34, 122.34, 118.64, 72.09, 72.09, 143.39, 180.34, 143.09, 98.09, 180.34, 180.34, 72.09, 143.39, 180.34, 72.09], 'NY': [146.69, 122.34, 118.64, 146.69, 122.34]}
	Contact State Profit 

#### Grouping 'Contact sex'

In [462]:
def get_range_date(sales_data_column):
    # Counting unique elements
    sales_data_category = Counter(sales_data_column)
    # Rounding numbers down to their nearest lower multiples (10, 20, 30, ...)
    rng = list(floor(k/10) * 10 for k in sales_data_category.keys())
    
    # All possible intervals
    possible_rng = range(min(rng), max(rng)+10, 10)
    # Dictionary of possible intervals
    rng_dict = {f"{i}-{i+10}": 0 for i in possible_rng}
    
    # Binning values into intervals of 10
    for i in possible_rng:
        for j in sales_data_category:
            if j >= i and j < i+10:
                rng_dict[f"{i}-{i+10}"] += sales_data_category[j]
    
    # Removing the count < 10
    return {rng: val for rng, val in rng_dict.items() if val > 0}

get_range_date(sales_data["Contact Age"])

{'20-30': 4, '30-40': 5, '40-50': 15, '50-60': 15}

In [463]:
def get_range_date_optimized(sales_data_column):
    
    sales_data_category = Counter(sales_data_column)

    range_dict = {}
    for value, count in sales_data_category.items():

        floored_value = floor(value/10) * 10 
        range_key = f"{floored_value}-{floored_value + 10}"

        if range_key in range_dict:
            range_dict[range_key] += count
        else:
            range_dict[range_key] = count

    return range_dict

get_range_date_optimized(sales_data["Contact Age"])

{'40-50': 15, '30-40': 5, '20-30': 4, '50-60': 15}

In [464]:
def get_range_profit_list(sales_data_column):
    
    sales_data_category = get_key_profit_dict(sales_data_column)

    range_dict = {}
    for value, count in sales_data_category.items():

        floored_value = floor(value/10) * 10 
        range_key = f"{floored_value}-{floored_value + 10}"

        if range_key in range_dict:
            range_dict[range_key] += count
        else:
            range_dict[range_key] = count

    return range_dict

get_range_profit_list(sales_data["Contact Age"])

{'40-50': [143.39,
  122.34,
  230.89,
  122.34,
  122.34,
  180.34,
  98.09,
  146.69,
  118.64,
  98.09,
  146.69,
  180.34,
  72.09,
  180.34,
  72.09],
 '30-40': [230.89, 143.09, 72.09, 98.09, 72.09],
 '20-30': [118.64, 146.69, 72.09, 143.09],
 '50-60': [98.09,
  180.34,
  122.34,
  143.09,
  143.39,
  146.69,
  143.39,
  230.89,
  230.89,
  122.34,
  143.39,
  143.39,
  118.64,
  180.34,
  180.34]}

In [465]:
def get_range_profit_mean_category_dict(sales_data_category):
    range_profit_dict =  get_range_profit_list(sales_data_category)
    return {k: custom_mean(v) for k, v in range_profit_dict.items()}

get_range_profit_mean_category_dict(sales_data["Contact Age"])

{'40-50': 135.64666666666668,
 '30-40': 123.25,
 '20-30': 120.1275,
 '50-60': 155.17}

In [466]:
for col in target_columns:
    print(f"{col} Data:")
    print(f"\t{dict(Counter(sales_data[col]))}")
    print(f"\t{col} Profit List: {get_key_profit_dict(sales_data[col])}")
    print(f"\t{col} Profit Mean: {get_profit_stat_category_dict(sales_data[col], stat="mean")}")
    print(f"\t{col} Standard Deviation: {get_profit_stat_category_dict(sales_data[col], stat="stdev")}")
    print(f"\t{col} Coefficient Variation: {get_profit_stat_category_dict(sales_data[col], stat="cv")}\n")

print(
    dict(sorted(get_range_date_optimized(sales_data["Contact Age"]).items(),
                key = lambda x: x[0]))
)
print(
    dict(sorted(get_range_profit_list(sales_data["Contact Age"]).items(),
                key = lambda x: x[0]))
)
print(
    dict(sorted(get_range_profit_mean_category_dict(sales_data["Contact Age"]).items(),
                key = lambda x: x[0]))
)

Contact Sex Data:
	{'M': 22, 'F': 17}
	Contact Sex Profit List: {'M': [143.39, 118.64, 72.09, 146.69, 122.34, 143.39, 180.34, 122.34, 118.64, 146.69, 72.09, 143.39, 180.34, 122.34, 143.09, 143.09, 98.09, 143.39, 180.34, 143.39, 180.34, 72.09], 'F': [230.89, 98.09, 230.89, 180.34, 230.89, 72.09, 98.09, 146.69, 122.34, 143.09, 118.64, 180.34, 230.89, 72.09, 98.09, 146.69, 122.34]}
	Contact Sex Profit Mean: {'M': 133.48, 'F': 148.38}
	Contact Sex Standard Deviation: {'M': 33.22, 'F': 56.33}
	Contact Sex Coefficient Variation: {'M': 0.25, 'F': 0.38}

Contact State Data:
	{'OH': 11, 'WV': 4, 'PA': 19, 'NY': 5}
	Contact State Profit List: {'OH': [143.39, 180.34, 146.69, 122.34, 230.89, 146.69, 98.09, 122.34, 143.09, 230.89, 98.09], 'WV': [230.89, 143.39, 143.09, 143.39], 'PA': [118.64, 72.09, 98.09, 230.89, 180.34, 122.34, 118.64, 72.09, 72.09, 143.39, 180.34, 143.09, 98.09, 180.34, 180.34, 72.09, 143.39, 180.34, 72.09], 'NY': [146.69, 122.34, 118.64, 146.69, 122.34]}
	Contact State Profit M

- In conclusion the porfitable age range is between 50 and 60 (highest average profit of 155.17)